### ISSUE TITLE
Datatype breaks get-model

### ISSUE THREAD

**Issue by vedadux** [Z3Prover/z3#2063]
I have used the z3 python API to try and generate a simple "next state logic" for a small circuit, for which I want to specify the behaviour. I declared a datatype `State` and said that for any possible current state and input, there must be a next state. The next state logic was modeled as an uninterpreted function that returns true for valid transitions. I know that I still have to create the condition that only one following state is allowed and so on, but even this breaks the model extraction:
```
; benchmark generated from python API
(set-info :status unknown)
(declare-datatypes ((State 0)) ((State (s0) (s1) (s2))))
 (declare-fun nsl (State Bool Bool State) Bool)
(assert
 (forall ((x State) (r1 Bool) (r2 Bool) )(exists ((y State) )(nsl x r1 r2 y))
 )
 )
(check-sat)
(get-model)
```
The assertion is found to be satisfiable, but something breaks when the (get-model) part is executed:
```
sat
(error "line 10 column 10: Sort mismatch at argument #1 for function (declare-fun = (Bool Bool) Bool) supplied sort is State")
```
The above formula was produced by the following python code:
``` python3
import z3
solver = z3.Solver()
B = z3.BoolSort()
S, (s0, s1, s2) = z3.EnumSort("State", ("s0", "s1", "s2"))
nsl = z3.Function("nsl", S, B, B, S, B)
x, y = z3.Consts("x y", S)
r1, r2 = z3.Consts("r1 r2", B)
nsc1 = z3.ForAll([x, r1, r2], z3.Exists([y,], nsl(x, r1, r2, y)))
solver.add(nsc1)
print(solver.to_smt2())
```
Calling `solver.check()` from python gives me sat, but calling `solver.model()` afterwards throws an exception, similar to the message, and says that the model is not available.

**Comment by vedadux** [comment]
Hi, this seems to also happen with builtin datatypes, such as Int. I have then tried to change all datatypes to Int and it seems to work and produces a model. It seems to be the mixing of datatypes in function calls that causes this.

**Comment by NikolajBjorner** [comment]
is this the current version of Z3? It doesn't reproduce in the version I am building.
The error message from SMTLIB is definitely a bug, but perhaps fixed. 
The exception from python may be because Z3 isn't able to determine whether the formula is sat.
If solver.check() returns unknown, the call to solver.model() will fail. That said, the current version of Z3 seems to produce "sat" when invoking solver.check() and there is a model as well.

**Comment by vedadux** [comment]
I have called `solver.check()`, which returned `sat`, and then tried extracting the model.
This was with the version 4.8.0 for 64 bit. As far as I see, the version available from pip is 4.8.0.0.post1 where this occurs.

I have tried it again on my work notebook which has version 4.5.1 and it works fine there. Very strange.

**Comment by NikolajBjorner** [comment]
not too strange: the algebraic datatypes were reworked around the point where your python installation was created. Bugs were introduced, then fixed. I don't manage the pip versions of Z3. Can you install one of the official releases instead for your work?
